kable-core/src/androidMain/kotlin/ThreadingStrategy.kt

```diff
@@ -0,0 +1,92 @@
+package com.juul.kable
+
+import com.juul.kable.OnDemandThreadingStrategy.acquire
+import com.juul.kable.OnDemandThreadingStrategy.release
+import kotlinx.atomicfu.atomic
+import kotlinx.atomicfu.locks.reentrantLock
+import kotlinx.atomicfu.locks.withLock
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlin.coroutines.cancellation.CancellationException
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.minutes
+import kotlin.time.TimeMark
+import kotlin.time.TimeSource
+
+private val threadNumber = atomic(0)
+private fun generateThreadName() = "Kable#${threadNumber.incrementAndGet()}"
+
+public interface ThreadingStrategy {
+    public fun acquire(): Threading
+    public fun release(threading: Threading)
+}
+
+/**
+ * A [ThreadingStrategy] that creates ["threads"][Threading] when [acquired][acquire] and
+ * immediately shuts down when [released][release].
+ */
+public object OnDemandThreadingStrategy : ThreadingStrategy {
+
+    override fun acquire(): Threading = Threading(generateThreadName())
+
+    override fun release(threading: Threading) {
+        threading.shutdown()
+    }
+}
+
+/**
+ * A [ThreadingStrategy] that pools unused ["threads"][Threading] until [evictAfter] time has
+ * elapsed.
+ *
+ * In most circumstances, only a a single [PooledThreadingStrategy] instance should be created per
+ * application run, as it holds the "shared" pool of unused ["threads"][Threading].
+ *
+ * Useful for when [Peripheral] connections are quickly being spun down and up again — as they can
+ * re-use existing ["threads"][Threading] ([acquire] their ["threads"][Threading] from the unused
+ * pool).
+ *
+ * If [Peripheral] connections are expected to be long running, or for there to be long down times
+ * between connections, [OnDemandThreadingStrategy] may be a better choice.
+ */
+public class PooledThreadingStrategy(
+    scope: CoroutineScope = GlobalScope,
+    private val evictAfter: Duration = 1.minutes,
+) : ThreadingStrategy {
+
+    private val pool = mutableListOf<Pair<TimeMark, Threading>>()
+    private val guard = reentrantLock()
+
+    private val job = scope.launch {
+        try {
+            while (true) {
+                guard.withLock {
+                    pool.removeAll { (timeMark) -> timeMark.hasPassedNow() }
+                }
+                delay(evictAfter / 2)
```

Comment:
Why half the eviction time? Sqrt might be a little more flexible.

Reply:
Was trying to find a balance between checking too often and not having threads stay alive for a lot longer than needed.

I do like `sqrt` for this, but unfortunately `evictAfter` is a `Duration`; whereas `div` is provided, but a `Duration.sqrt` function is not. I could convert to milliseconds for all durations, do the `sqrt` operation, then convert back, but that seems like an unnecessary amount of complexity for little gain?

Perhaps dividing by `3` would be better?